src/metabase/models/audit_log.clj

```diff
@@ -31,51 +34,73 @@
-                             (:raw? object) (dissoc object :raw?)
-                             (t2/model object) (model-details object unqualified-topic)
-                             :else (or object {}))]
-     (t2/insert! :model/AuditLog
-                 :topic    unqualified-topic
-                 :details  details
-                 :model    model-name
-                 :model_id model-id
-                 :user_id  user-id)
-     ;; TODO: temporarily double-writing to the `activity` table, delete this in Metabase v48
-     (when-not (#{:card-read :dashboard-read :table-read :card-query :setting-update} unqualified-topic)
+  `:object` and `:previous-object` both have `model-details` called on them to determine which fields should be audited,
+  then they are added to `:details` before the event is recorded. `:previous-object` is only included if any audited fields
+  were updated."
+  [topic :- :keyword
+   params :- [:map {:closed true}
+              [:object          {:optional true} [:maybe :map]]
+              [:previous-object {:optional true} [:maybe :map]]
+              [:user-id         {:optional true} [:maybe pos-int?]]
+              [:model           {:optional true} [:maybe [:or :keyword :string]]]
+              [:model-id        {:optional true} [:maybe pos-int?]]
+              [:details         {:optional true} [:maybe :map]]]]
+  (let [unqualified-topic (keyword (name topic))
+        object            (:object params)
+        previous-object   (:previous-object params)
+        object-details    (model-details object unqualified-topic)
+        previous-details  (model-details previous-object unqualified-topic)
+        user-id           (or (:user-id params) api/*current-user-id*)
+        model             (model-name (or (:model params) object))
+        model-id          (or (:model-id params) (u/id object))
+        details           (merge {}
+                           (:details params)
+                           (if (not-empty previous-object)
+                             (prepare-update-event-data object-details previous-details)
+                             object-details))]
+    (t2/insert! :model/AuditLog
+                :topic    unqualified-topic
+                :details  details
+                :model    model
+                :model_id model-id
+                :user_id  user-id)
+    ;; TODO: temporarily double-writing to the `activity` table, delete this in Metabase v48
+    ;; TODO figure out set of events to actually continue recording in activity
+    (when-not (#{:card-read :dashboard-read :table-read :card-query :setting-update} unqualified-topic)
       (activity/record-activity!
-       :topic      topic
-       :object     object
-       :details-fn (fn [_] details)
-       :user-id    user-id)))))
+       {:topic    topic
+        :object   object
+        :details  details
+        :model    model
+        :model-id model-id
+        :user-id  user-id}))))
```

Comment:
I'd like to propose separating the generation of the value on 98:103 from actually doing the side effect. 

So, instead of:

```clojure
(record-event! topic params)
```

we would call:

```clojure
(actually-record-the-event! 
  (generate-event topic params))
```

----

## Why? 

It makes it easier to see and test what the outputs for record-object! are. Yes, now we can use the database to see, but it would makes it easier to see them in the repl, too. The tests that check it could be run in parallel, and would be way faster, because they don't rely on the db.

It feels like such a nice spot to separate the imperative shell from the complicated pure logic-y bit.

wdyt?

Reply:
In fact we could have a long test using e.g. `are` with 50 inputs/outputs and a simple `=` check for `generate-event` that is simple, and easy to follow (v.s. doing database setup/teardown/etc. for each one).